Userland/Libraries/LibJS/Parser.cpp

```diff
@@ -241,12 +267,98 @@ class ScopePusher {
             }
         }
 
+        for (auto& it : m_identifier_groups) {
+            auto const& identifier_group_name = it.key;
+            auto& identifier_group = it.value;
+
+            if (identifier_group_name == "arguments"sv) {
+                // NOTE: arguments is a special variable that should not be treated as a candidate to become local
+                continue;
+            }
+
+            bool scope_has_declaration = false;
+            MUST(m_node->for_each_var_declared_name([&](auto const& name) {
+                if (m_function_parameters.has_value() && m_forbidden_lexical_names.contains(name))
+                    return;
+                if (name == identifier_group_name)
+                    scope_has_declaration = true;
+            }));
+            MUST(m_node->for_each_lexically_declared_name([&](auto const& name) {
+                if (name == identifier_group_name)
+                    scope_has_declaration = true;
+            }));
```

Comment:
It would be nice to break out of these loops immediately when `scope_has_declaration` becomes `true`.